### ISSUE TITLE
Curriculum: Projects that require user to dive into existing codebases

### ISSUE THREAD

**Issue by nikitarevenco** [TheOdinProject/curriculum#27553]
### Checks

- [X] This is not a duplicate of an existing issue (please have a look through our [open issues list](https://github.com/TheOdinProject/curriculum/issues) to make sure)
- [X] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/.github/blob/main/CONTRIBUTING.md)
- [X] Would you like to work on this issue?

### Describe your suggestion

I think it would be a good idea if there were projects which don't have the user implement an app from scratch - but rather have an app that is already working and add features to it.

For example, there could be a project in the JavaScript section that is a snake game with some basic features like the snake grows, the snake can eat food, food appears randomly etc. And some of the points of the assignment would be for example
- have the snake wrap around the map if it reaches the edge of one side
- introduce different foods that give the snake more points when eaten
- introduce hazards and enemies that will try to follow the snake and hurt it 

Just a silly little example. I think this would be useful because:
- More often than not, people will have to work on existing code rather than start from scratch.
- It allows learners to develop skills to work with other people's code, which may use different approaches to their own
- Builds skills to read code. If the project at the beginning is complicated and has a lot of things in it, for example a project later in the React course could have several dozen files and the learner would have to dive into the code base and understand the code before starting to work on the assignment. 

This could be introduced into the curriculum by adding new projects, or replacing existing projects with this new type of project. 
Let me know what you think.

### Path

Foundations, Ruby / Rails, Node / JS

### Lesson Url

Curriculum

### (Optional) Discord Name

Revenco

**Comment by JoshDevHub** [comment]
I think this kind of thing is potentially a cool idea, and I've discussed this internally with a couple of maintainers before.

Another nice thing to potentially explore with this idea: intentionally write a bug or two into the existing app. Then learners could be told fix the problems in the app. This is a *very* common type of real world task that's a bit different from feature work.

To me, one of the main cons with this type of thing is that it probably doesn't play well on a portfolio? It's more of a theoretical learning exercise, which is fine. I just think that potentially by the time learners are advanced enough to work on this kind of thing, they're going to be getting close to job ready and will want to spend their time on projects that can go on a resume. But maybe that's just a con in my mind.

Curious if other maintainers have thoughts on this kind of thing @TheOdinProject/maintainers

**Comment by cakegod** [comment]
I believe it is a good idea. These exercises have the potential to improve code comprehension and debugging skills. Not everything needs to be explicitly included in the portfolio.

We can incorporate these exercises throughout the curriculum with varying levels of difficulty based on placement in the curriculum.

**Comment by nikitarevenco** [comment]
If people are on board with this, I think a good place to get started would be the JavaScript Foundations section. Specifically, after the "handling javascript error" lesson. 

#### What the user will get when they fork the repo

The assignment could be something like this:
#### Fix bugs
- The game doesn't even start. Find the error and fix it so that it works! 
- Inputting in something other than "higher", "lower" or "null" will stop the program instead of going onto the next question. Can you try figure out why?
- Clicking on "cancel" doesn't work - fix it!
- The program incorrectly guesses the number. 
#### Features
- Add a feature so when play() is called, the user is prompted for for difficulty - either "Easy", "Medium" or "Hard" which would change the maximum values from 10, 100 and 1000.
- Make the game playable in reverse! By passing in play(reverse) the player should now be the one guessing, and the computer telling the player whether they their guess was higher, lower or correct

```javascript
function play() {
    let min = 1;
    let max = 100;
    let guess = Math.floor((max + min) / 2);

    for (let i = 0; i++; i < 30) {
        if (guess === max || guess === min) {
            alert(`Your number must be ${guess}.`);
            break;
        }
        let userInput = prompt(`My guess: ${guess}. If your number is higher, say "higher". If it's lower, say "lower". If it's correct, say "correct".`);

        if (userInput === "correct") {
            alert(`I won! Your number was ${guess}.`);
            break;
        } else if (userInput === "higher") {
            min = guess;
        } else if (userInput === "lower") {
            max = guess;
        } else if (userInput === "null") { 
            alert("You quit the game!");
            break;
        } else {
            alert("Invalid input! Try again.");
            break;
        }
        guess = Math.floor((min + max) / 2);
    }
}
```

**Comment by MaoShizhong** [comment]
I like the idea for sure and agree that not everything *needs* to be portfolio-focused. There are already plenty of great portfolio candidates around the areas where people seem to start applying for jobs.

I'm not so sure about introducing such kinds of exercises so early as Foundations though. I agree with Josh in that there probably should be a little more experience with reading and writing code than you'd likely expect people in Foundations to have. Plus, in Foundations, we already have some "fix these bugs" beginner exercises before the repo/Jest-based exercises later.

Two prime spots for me are towards the end of the JS section and towards the end of React section.
Similar to the way the CS section is handled between pathways, perhaps then a Ruby equivalent towards the end of that course and the JS section one omitted as "a duplicate".

I really like the idea of having more in the React section overall. A "fix this React app and add XYZ features to it" project/exercise before the "Shopping Cart capstone" I think might be both fun and also break up the slog of reading about routers, fetching *and* testing between Memory Card and Shopping Cart.

I know quite a few people in the community express surprise at the end of the React section going "is that it?". If they're taking the Node pathway then it'll come back in the latter part of Node but they don't necessarily know that. But if they're taking the Rails pathway, then that really is it.

**Comment by thatblindgeye** [comment]
This sounds like a really nice idea! I can definitely see the merit in helping expose users to reading code someone else wrote and having to fix or add things to it, rather than doing everything from scratch.

Agreed that Foundations may not be the best spot, but this is worth more of a discussion. Depending on the implementation, we could always introduce some of this stuff in the JS exercise repo, either instead of or in addition to a project.

Maybe starting out with more simple exercises to get users used to reading smaller blocks of code, before eventually ramping up to a larger project that requires users to have to traverse multiple files and such?

No actual points but @nikitarevenco ++ for this idea! (and I guess @JoshDevHub ++ too 😆 )

**Comment by CouchofTomato** [comment]
@nikitarevenco Are you still working on this?

**Comment by thatblindgeye** [comment]
This might be better served converting to a discussion for now since there's nothing actionable yet to do. It may require further nailing down if ultimately it's something we want to move forward with, where it would live if so, and how exactly to implement it.

**Comment by nikitarevenco** [comment]
> This might be better served converting to a discussion for now since there's nothing actionable yet to do. It may require further nailing down if ultimately it's something we want to move forward with, where it would live if so, and how exactly to implement it.

I love Mao's idea to add a project at the end of the react course that is exactly like this. Once I completed it I was like 'So this is it?', would be nice to have some sort of final project to sum up the previous experience and have a project that is closer to something you'd encounter in the real world.

Maybe to get started we can begin brainstorming ideas for what the project could be, what the learner would have to do etc? I think it will be way easier to focus on a single project like this for now to get the ball rolling.

I think that the project could give learner a fully working e-commerce site (as far as we can go - so probably using local storage for database and no payment system) that is written in raw javascript and has to be converted to use React, or maybe it has a separate HTML document for each page and the learner needs to convert it to a single page application. Alternatively, they may be tasked with solving several smaller issues. Like there could be 5 issues eg fix this bug, or add that feature. 

Just throwing a bunch of ideas together and maybe something will stick
